File: JeongGod/18week/baekjoon/B10775.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] root;
    public static int find(int p) {
        if (root[p] == p) {
            return p;
        }
        root[p] = find(root[p]);
        return root[p];
    }

    public static void union(int child, int parent) {
        root[child] = root[parent];
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st;

        int G, P;

        G = Integer.parseInt(br.readLine());
        P = Integer.parseInt(br.readLine());

        int[] airplains = new int[P];
        root = new int[G+1];
        for (int i=0; i<=G; i++) {
            root[i] = i;
        }

        int answer = 0;
        for (int i=0; i<P; i++) {
            int gate = Integer.parseInt(br.readLine());
            // 만약 부모를 찾았는데 0이면 더 이상 도킹할 수 없다는 얘기
            int parent = find(gate);
            if (parent == 0) break;
            answer++;
            union(parent, parent-1);
        }
        System.out.println(answer);

    }
}

File: jekyun-park/week18/boj/b10775.swift
//
//  main.swift
//  SwiftAlgorithms
//
//  Created by 박제균 on 2022/05/02.
//  BOJ > 10775 > 공항

import Foundation

let G = Int(readLine()!)!
let P = Int(readLine()!)!
var gates = [Int]()
var parent = Array(0...G)

func find(_ n: Int) -> Int {
    if n == parent[n] { return n }
    return find(parent[n])
}

func union(_ a: Int, _ b: Int) {
    parent[a] = parent[b]
}

var answer = 0

for _ in 0..<P {
    gates.append(Int(readLine()!)!)
}

for i in gates {
    let parent = find(i)
    if parent == 0 { break }
    answer += 1
    union(parent, parent-1)
}

print(answer)
